### DISCUSSION TITLE
KeyUsageConstraints configuration on validate_pdf_signature

### DISCUSSION THREAD

**Discussion by inui-sdt2** [MatthiasValvekens/pyHanko#59] [Q&A] [answered]
this is from https://github.com/MatthiasValvekens/pyHanko/issues/10#issuecomment-974778517

Dear @MatthiasValvekens  and all
thank you for your previous support.
I haven't been able to get it to work yet, so could you give me some advice?

the actual signing certificate contains 'digital_signature' and 'key_encipherment
so I created the following code for signed pdf validation.

```
vc = ValidationContext(
    trust_roots=[root_cert],
    other_certs=[int_cert]
    )

ts_vc = ValidationContext(
    trust_roots=[tsa_root_cert]
    )   

key_usage_settings = KeyUsageConstraints(
    key_usage={'digital_signature', 'key_encipherment'}, match_all_key_usages=False
)

with open('signed-14.pdf', 'rb') as doc:
    r = PdfFileReader(doc)
    sig = r.embedded_signatures[0]
    status = validate_pdf_signature(sig, vc, ts_vc, key_usage_settings)
    print(status.pretty_print_details())

```

But I am still encountering  the error as below.

```
The active key usage policy requires at least one of the key usage extensions non repudiation to be present.
Chain of trust validation for Common Name: xxxx failed.
Signer info
-----------
Certificate subject: "Common Name:xxxxx"
Certificate SHA1 fingerprint: 023cdf2b77e06809c1590484888b4c1b7dfcdxxxx
Certificate SHA256 fingerprint: 55b5636fcd2da13b137506d957e6c89a270c9c391bd78f290f4518ddfd4xxxx
Trust anchor: "No path to trust anchor found."
The signer's certificate is untrusted.
```

**Comment by MatthiasValvekens** [comment, marked as answer]
Hi @inui-sdt2, thanks for moving this conversation to the discussion board.

The following line stands out to me:

```python
    status = validate_pdf_signature(sig, vc, ts_vc, key_usage_settings)
```

As it happens, the fourth positional argument to `validate_pdf_signature` is actually `diff_policy`, not `key_usage_settings`. 
Could you try replacing that line with this one?

```python
    status = validate_pdf_signature(sig, vc, ts_vc, key_usage_settings=key_usage_settings)
```

In hindsight, I should probably have made those policy description arguments keyword-only, then you would've gotten a runtime error. In this case, since your document presumably only contains one signature, the `diff_policy` was never used, so it went by unnoticed.


------

EDIT: By the way, adding `'key_encipherment'` to the `KeyUsageConstraints` settings is probably not what you want, especially if you have `match_all_key_usages` set to `False`. After all, you wouldn't want to accept a certificate that was only appropriate for key encipherment but not digital signing!

**Reply by inui-sdt2** [reply]
Hi Matthias
It works perfectly with your advice! thank you.
(and thanks for additional information. this helps my future developments)
I will continue to play this cool name tool!

**Reply by MatthiasValvekens** [reply]
Great! Thanks for letting me know :)

**Reply by gavan24** [reply]
Hi there,
is it possible to switch of "Chain of trust validation"? I just want to verify if a signature it present or not and print out common name email address and country. I also keep getting: "Chain of trust validation for Common Name: xxxx failed."
Can anyone help there?
Thanks a million,
Gavan

**Reply by MatthiasValvekens** [reply]
Hi @gavan24, if you're only interested in the metadata of the signature, then you don't need to call the validation routine at all. `EmbeddedPdfSignature` has a `signer_cert` property that you can use to retrieve the certificate: https://github.com/MatthiasValvekens/pyHanko/blob/b26c28c7c9d9110a72582ee0e941b95a6c67d7bc/pyhanko/sign/validation/pdf_embedded.py#L228-L234.

If you absolutely want to run the validation logic while disabling trust validation specifically (you probably don't), then passing `signer_validation_context=None` in `validate_pdf_signature` would work.
